### PR TITLE
Add & Test Clang Tidy Review Script

### LINKED ISSUE #13: Conclude the handling of Clang-Format and Clang-Tidy
Refering to issue https://github.com/BoostGSoC20/ublas/issues/4#issuecomment-645274452
We need to conclude whether we want to have a review script and github action, see branch https://github.com/BoostGSoC20/ublas/tree/clang-tidy-review-check

### PR DESCRIPTION
**Pull request by coder3101**

### CODE REVIEW

**Comment by coder3101** [comment]
@bassoy, I have written the review script and now I am testing it here 🤞 

You can see above that this PR passed all checks because this PR did not actually change any cpp/hpp file at that time.

**Comment by coder3101** [comment]
Checking if the bot will spam when PR is synchronized! 🤞

**Comment by coder3101** [comment]
Bot is spamming 😞 

I will try to fix it tomorrow.

File: .ci/post_review.py
Comment (coder3101) [inline .ci/post_review.py:158]: Due to some reason, this line is not being invoked!

File: .ci/post_review.py
Comment (coder3101) [inline .ci/post_review.py]:
```suggestion
        if target in reports:
```

Eureka! found it. Will continue the work next day.

File: .ci/post_review.py
Comment (coder3101) [inline .ci/post_review.py]:
```suggestion
        comment_lgtm(pr_handle)
```

**Comment by coder3101** [comment]
Okay, So everything seems fine by now! Let's see **if bot only reviews the newly changed files** or not. For simplicity, I will format a file with clang-format.

**Comment by coder3101** [comment]
`strides.hpp` did not have any clang-tidy issues or warnings. I will format another file `fixed_rank_stride.hpp` which does have issues to check if they are reported by clang-tidy bot.

**Comment by coder3101** [comment]
As this PR was synchronized, only the new changes have been reported by the Bot as you can see above.

Now, I will revert all the code formattings, It will have code just for Clang tidy review which @bassoy should review. I also request you to please approve the PR when you find it acceptable.

**Comment by coder3101** [comment]
@bassoy Please review and approve the PR

File: .github/workflows/clangtidy_review.yml
Comment (bassoy) [inline .github/workflows/clangtidy_review.yml]: here you are install libboost from the ubuntu 18.04 packages.
is that wanted? we should clone the repo and test either the development or master branch

File: .github/workflows/clangtidy_review.yml
Comment (bassoy) [inline .github/workflows/clangtidy_review.yml]: Let's run it on ubuntu 20.04

File: .ci/post_review.py
Comment (bassoy) [inline .ci/post_review.py]: I think it is sufficient to mention the license file as we did for other source code files:
```py
# Distributed under the Boost Software License, Version 1.0. (See
# accompanying file LICENSE_1_0.txt or copy at
# http://www.boost.org/LICENSE_1_0.txt)
```

File: .ci/post_review.py
Comment (bassoy) [inline .ci/post_review.py]: Suggestion to put the date of publication.
```py
# Script for automatically creating review on github from clang-tidy report
# Copyright (c) Ashar Khan <ashar786khan.at.gmail.com> 2020
```

File: .ci/post_review.py
Comment (bassoy) [inline .ci/post_review.py]: I am not a python expert, but are the line breaks wanted?

**Review by bassoy — CHANGES_REQUESTED** [review]
Good work. Few minor marks.
Please review my comments and answer them.

**Comment by bassoy** [comment]
You have many checks queued. Can you cancel those which are not necessary?

File: .ci/post_review.py
Comment (coder3101) [reply .ci/post_review.py]:
```suggestion
# Copyright (c) 2020 Ashar Khan <ashar786khan.at.gmail.com>
```

File: .ci/post_review.py
Comment (coder3101) [reply .ci/post_review.py]:
```suggestion
```

Agreed! I have removed the Extra License Text

File: .ci/post_review.py
Comment (coder3101) [reply .ci/post_review.py]:
```suggestion
            abs_file_path, line_number, _, severity, message = content[i].split(
                ":", maxsplit=4)
            relative_path = os.path.relpath(abs_file_path, root_path)

            k = 1
            body = ""
            while i+k < n and not ("error" in content[i+k] or "warning" in content[i+k]):
                body += "\n" + content[i + k].replace(abs_file_path, relative_path)
                k += 1
            i += k

```

I ran VSCode Format document for this file and it made this change.

File: .github/workflows/clangtidy_review.yml
Comment (coder3101) [reply .github/workflows/clangtidy_review.yml]: The bare basic minimum for `clang-tidy` is that it finds all includes in codes it is running on, Since we use `boost/config` and other boost libs, it is required that clang-tidy finds all these includes.

Now, there are two ways for it:
 - Setup Boost ROOT or
 - Use `libboost-all-dev` so that all Boost headers are found in the system includes.

We then run the clang-tidy checks with `-Iinclude` flag, this makes sure that we are not testing on system boost.ublas rather in our repository. 

This is quicker than setting up boost root since we don't care about clang-tidy warnings from other Boost libraries, we can use any boost version. As for the build test, We are running with boostorg/boost from the master.

File: .ci/post_review.py
Comment (coder3101) [inline .ci/post_review.py]:
```suggestion
Licensed under Boost Software License 1.0
```

File: .github/workflows/clangtidy_review.yml
Comment (coder3101) [reply .github/workflows/clangtidy_review.yml]: From Gitter chat:

> I have upgraded the machines to use ubuntu 20.04 on all Linux instances except for clang-tidy check Job, it still uses ubuntu 18.04 but uses clang-tidy-10. I could not upgrade it because on ubuntu 20.04 the multi-precision library does not compile with clang-10. Maybe in future when new boost releases are made, we will upgrade it to 20.04. See this, I tried to upgrade but got [this error](https://github.com/BoostGSoC20/ublas/runs/775558213?check_suite_focus=true#step:5:65)

We do use multi-precision in one of the Unit Tests! It is that error.

**Comment by coder3101** [comment]
> 
> 
> You have many checks queued. Can you cancel those which are not necessary?


Sure

File: .github/workflows/clangtidy_review.yml
Comment (coder3101) [reply .github/workflows/clangtidy_review.yml]: Does it really matter what boost library we use, as long as we are able to lint our repository with any boost version. In Builds and Sanitizer checks, We do use Boost from master.

If you want, I can setup boost from master as well. But that will increase time for checks as everytime it will need to compile and install Boost from source.

**Comment by coder3101** [comment]
After this work is merged. I will take some time and try to implement `ccache` for caching our build artifacts for a faster build times.

File: .github/workflows/clangtidy_review.yml
Comment (bassoy) [reply .github/workflows/clangtidy_review.yml]: Yes. You are right. The reason why I included multiprecision is that I wanted to make sure that we can have also more complex types supporting our library. It is not that crucial. we could also think of including `std::ratio` or `std::complex` instead.
Hence my suggestion: comment out multiprecision and include `std::complex` and **use ubuntu 20.04** and simplify the installation and usage of later clang and gcc versions.

File: .github/workflows/clangtidy_review.yml
Comment (bassoy) [reply .github/workflows/clangtidy_review.yml]: > Does it really matter what boost library we use, as long as we are able to lint our repository with any boost version. In Builds and Sanitizer checks, We do use Boost from master.

Generally, I am for your approach. Less work! However, the boost-package in ubuntu-20.04 could be version1.73 while the master on github could be already 1.75 and all of a sudden we get a compile error.

**Comment by bassoy** [comment]
> After this work is merged. I will take some time and try to implement `ccache` for caching our build artifacts for a faster build times.

Very good. However, I think our biggest concern with our CI-time is the queuing time.

File: .github/workflows/clangtidy_review.yml
Comment (coder3101) [reply .github/workflows/clangtidy_review.yml]: Okay, I will build boost from source master and compile and install it. I will check if it's possible to cache it as well, so we can speed up things a bit.

File: .github/workflows/clangtidy_review.yml
Comment (coder3101) [reply .github/workflows/clangtidy_review.yml]: Okay, Got it!

**Comment by coder3101** [comment]
> > After this work is merged. I will take some time and try to implement `ccache` for caching our build artifacts for a faster build times.
> 
> Very good. However, I think our biggest concern with our CI-time is the queuing time.

The Queue time is only for this PR, because it is from same repository, every push to this PR is being built twice (one for Push and Another for Pull), In General Push or Pull request from other contributors. Will not have this much tests. 

There is another way to significantly decrease the Queue by 'fail-fast' strategy. If any one build test fails the complete queue is cancelled. Currently it is set to false, so even if clang format fails the jobs run on gcc/Linux etc. 

We can turn fail fast to true, in future.

File: .github/workflows/clangtidy_review.yml
Comment (coder3101) [reply .github/workflows/clangtidy_review.yml]: Best way would be to pull latest boost libraries as tarbar from web and install them. What do you say? It will be faster than compiling master from source, and latest as well.

File: .github/workflows/clangtidy_review.yml
Comment (coder3101) [reply .github/workflows/clangtidy_review.yml]: No, need. Seems like this issue has been fixed in `boost/multiprecision` master

File: .github/workflows/clangtidy_review.yml
Comment (coder3101) [reply .github/workflows/clangtidy_review.yml]: I am using 20.04 with `boostorg/master`. Since we only need headers files. I clone master and clone all submodules (shallow) then generate headers with `b2 headers` and manually copy them to `/usr/include`. while running clang-tidy.

**Comment by coder3101** [comment]
After I fixed the last issue the number of issues reported by clang-tidy decreased by one. This proved that we are indeed linting on our ublas repository, not on boostorg submodule clone of ublas.

@bassoy I request you to please review it now!

**Review by bassoy — APPROVED** [review]
The script produces a nice output. 
Please put comments on top of your functions in your python script.

**Comment by bassoy** [comment]
You can merge the pull request to github-actions branch. 
please comment your python script before you do so.

**Comment by coder3101** [comment]
I will do so and merge it tomorrow.

**Pull request merged**
